### DISCUSSION TITLE
gunicorn, threads, workers. How can I use the same object across all threads when using Flask-Socket

### DISCUSSION THREAD

**Discussion by noahra** [miguelgrinberg/Flask-SocketIO#1989] [General]
Hello,

I have been following this guide and modified the code to conform with my project:
https://blog.miguelgrinberg.com/post/add-a-websocket-route-to-your-flask-2-x-application

My goal is to create a web app where you can control a quadcopter-drone with the Crazyswarm library. 

Here is my  app.py

```
from flask import Flask, render_template
from flask_sock import Sock
from pycrazyswarm import Crazyswarm


swarm = Crazyswarm()
cf = swarm.allcfs.crazyflies[0]

app = Flask(_name_)
sock = Sock(app)

@app.route('/')
def index():
    return render_template('index.html')
@sock.route('/action')
def echo(sock):
    while True:
        data = sock.receive()
        if data == 'left':
            print("LEFT")
            cf.goto([-1.,0,0],0,0,True,1.)
        if data == 'right':
            cf.goto([1.,0,0],0,0,True,1.)
        if data == 'forward':
            cf.goto([0.,1.,0],0,0,True,1.)
        if data == 'back':
            cf.goto([0.,-1.,0],0,0,True,1.)
        if data =='take off':
            print('take off....')
            cf.takeoff(targetHeight=1.0,duration=2.5)
        if data =='land':
            print('...landing, please wait')
            cf.land(targetHeight=1.0,duration=2.5)
```
index.html:

```
<!doctype html>
<html>
  <head>
    <title>Drone</title>
  </head>
  <script>
     
    
     var socket = new WebSocket('wss://' + location.host + '/action');


     function sendText(button) {
     var text = button.value; // Get the value of the button
     socket.send(text); // Send the text to the server
    // socket.onopen = () => conn.send(text);
   }
   </script>
  <body>
    <h1>Drone Controller</h1>
    <div id="log"></div>
    <br>
    <button value="left" onclick="sendText(this)">Left</button>
  <button value="right" onclick="sendText(this)">Right</button>
  <button value="back" onclick="sendText(this)">Back</button>
  <button value="forward" onclick="sendText(this)">Forward</button>
  <button value="land" onclick="sendText(this)">Land</button>
  <button value="take off" onclick="sendText(this)">Take off</button>
   
  </body>
</html>
```

The issue lies within using this command 

`$ gunicorn -b 0.0.0.0:5000 --workers 4 --threads 100 app:app`

As multiple threads are run, this line in app.py is called multiple times:
`cf = swarm.allcfs.crazyflies[0]`

This creates a conflict, as this drone already has been instantiated in a different thread.

This leads to my program terminating.

If I only use one thread, the program will die rather quickly so I need the threads.

Is there a solution for only using on object across all threads? Because clearly, each thread instantiates that object..

I appreciate all answers, please ask for more details if needed. Sorry if I am being unclear.

Warm regards,
Noah

**Comment by miguelgrinberg** [comment]
I think you are confusing threads with processes here. You are running 4 processes, so your global variables will be created four times. Change the `--workers 4` to `--workers 1` and you'll have only one process. Within that process, it is okay to have 100 threads, all those share the same context and will use the same set of global variables.

**Reply by noahra** [reply]
Ok that is perfect! However! I still get the following error:

```
/home/user/code/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py:590: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  cfg = yaml.load(crazyflies_yaml)
Warning: No joystick found!
take off....
Exception in thread Thread-17:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.8/dist-packages/simple_websocket/ws.py", line 169, in _thread
    self.connected = self._handle_events()
  File "/usr/local/lib/python3.8/dist-packages/simple_websocket/ws.py", line 249, in _handle_events
    self.sock.send(out_data)
OSError: [Errno 9] Bad file descriptor
^C[2023-05-16 11:39:49 +0000] [76793] [INFO] Handling signal: int
[2023-05-16 11:39:50 +0000] [76825] [INFO] Worker exiting (pid: 76825)
[2023-05-16 11:39:50 +0000] [76793] [INFO] Shutting down: Master
```

I interpret `"OSError: [Errno 9] Bad file descriptor"` as me writing to a Websocket connection that is closed. However, I never close a connection anywhere in my app.py code?

**Reply by miguelgrinberg** [reply]
Maybe the client went away? That would also trigger an `OSError`. I would have expected a "broken pipe" error in that case, but "bad file descriptor" can also occur if you try to send something that the other side closed the connection.

**Reply by noahra** [reply]
Do you mean that somehow the frontend-part disconnects?

**Reply by miguelgrinberg** [reply]
Yes, that is what I mean.

**Reply by noahra** [reply]
Ok but then is it an issue related to my index.html code?

**Reply by miguelgrinberg** [reply]
I have no way to know that. The client may have closed the browser voluntarily, or maybe there are networking issues. If you suspect your client is closing a connection when it shouldn't, then you should debug your client and obtain more information.
